src/Datadog.Trace/Sampling/RateLimiter.cs

```diff
@@ -0,0 +1,155 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Sampling
+{
+    internal class RateLimiter : IRateLimiter
+    {
+        private static readonly Vendors.Serilog.ILogger Log = DatadogLogging.For<RateLimiter>();
+
+        private readonly ManualResetEventSlim _refreshEvent = new ManualResetEventSlim(initialState: false);
+
+        private readonly int _maxTracesPerInterval;
+        private readonly int _intervalMilliseconds;
+
+        private DateTime _windowBegin = DateTime.Now;
+        private DateTime _lastRefresh = DateTime.Now;
+        private ConcurrentQueue<ulong> _allowedTracesForInterval = new ConcurrentQueue<ulong>();
+
+        private int _windowChecks = 0;
+        private int _windowAllowed = 0;
+
+        private int _previousWindowChecks = 0;
+        private int _previousWindowAllowed = 0;
+
+        public RateLimiter(int? maxTracesPerInterval)
+        {
+            _maxTracesPerInterval = maxTracesPerInterval ?? 100;
+            _intervalMilliseconds = 1_000;
+            Log.Debug("Allowing {0} p1 traces per {1} milliseconds.", _maxTracesPerInterval, _intervalMilliseconds);
+        }
+
+        public bool Allowed(ulong traceId)
+        {
+            if (_maxTracesPerInterval == 0)
+            {
+                // Rate limit of 0 blocks everything
+                return false;
+            }
+
+            if (_maxTracesPerInterval < 0)
+            {
+                // Negative rate limit disables rate limiting
```

Comment:
Would a null `_maxTracesPerInterval` to mean "theres is no limit" make more sense?

Reply:
This is based on the Python implementation.
I'm happy to go with null, but I'd like to check to see if we're looking to standardize

Reply:
It makes sense standardize external-facing aspects like setting names (for example, all the language should use `DD_TRACE_DEBUG` and not something like `DATADOG_DEBUG_TRACE`). Actual implementations, however, should be idiomatic.

Reply:
Sounds good. Thanks!

Discussed offline: null is already used to mean "use the default setting", so we can't use it for "disable this feature". Maybe later we can add a second boolean setting for enable/disable, but not required for this PR.